File: apps/le-clap-expo/modules/leclap-ffmpeg/ios/LeclapFfmpegModule.swift
import ExpoModulesCore

// Bridges the embedded FFmpeg CLI engine (Rust + fftools, statically linked into
// libleclap_ffmpeg_core.a) to JS — the iOS counterpart of the Android Kotlin module. `run`/`probe`
// block for the whole job, so they run off the JS thread via AsyncFunction. The free functions
// `version()`, `run(args:)`, `probe(args:)` come from the uniffi-generated binding (Generated/).
public class LeclapFfmpegModule: Module {
  public func definition() -> ModuleDefinition {
    Name("LeclapFfmpeg")

    Function("version") {
      return version()
    }

    AsyncFunction("run") { (args: [String]) -> [String: Any] in
      let result = run(args: args)
      return ["code": result.code, "log": result.log]
    }

    AsyncFunction("probe") { (args: [String]) -> [String: Any] in
      let result = probe(args: args)
      return ["code": result.code, "output": result.output]
    }
  }
}

File: apps/le-clap-expo/modules/leclap-ffmpeg/android/src/main/java/expo/modules/leclapffmpeg/LeclapFfmpegModule.kt
package expo.modules.leclapffmpeg

import expo.modules.kotlin.Promise
import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition
import uniffi.leclap_ffmpeg_core.cancel as nativeCancel
import uniffi.leclap_ffmpeg_core.probe as nativeProbe
import uniffi.leclap_ffmpeg_core.run as nativeRun
import uniffi.leclap_ffmpeg_core.version as nativeVersion

/**
 * Bridges the embedded FFmpeg CLI engine (Rust + fftools) to JS. `run`/`probe` execute an ffmpeg /
 * ffprobe command; both block for the whole job so they run off the JS thread via AsyncFunction.
 */
class LeclapFfmpegModule : Module() {
  override fun definition() = ModuleDefinition {
    Name("LeclapFfmpeg")

    Function("version") { nativeVersion() }

    // Cooperative cancel of the in-flight run: sets the static shutdown flags (same effect as SIGTERM); returns before ffmpeg exits.
    Function("cancel") { nativeCancel() }

    AsyncFunction("run") { args: List<String>, promise: Promise ->
      try {
        val result = nativeRun(args)
        promise.resolve(mapOf("code" to result.code, "log" to result.log))
      } catch (e: Throwable) {
        promise.reject("ERR_RUN", e.message ?: "ffmpeg run failed", e)
      }
    }

    AsyncFunction("probe") { args: List<String>, promise: Promise ->
      try {
        val result = nativeProbe(args)
        promise.resolve(mapOf("code" to result.code, "output" to result.output))
      } catch (e: Throwable) {
        promise.reject("ERR_PROBE", e.message ?: "ffprobe failed", e)
      }
    }
  }
}
